segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableEntryIterator.java

```diff
@@ -0,0 +1,328 @@
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import io.pravega.segmentstore.server.reading.AsyncReadResultProcessor;
+import lombok.AccessLevel;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Iterates through {@link TableBucket}s in a Segment.
+ * @param <T> Type of the final, converted result.
+ */
+@Slf4j
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+@ThreadSafe
+class TableEntryIterator<T> implements AsyncIterator<T> {
+    //region Members
+
+    // The maximum size (in bytes) of each read to perform on the segment.
+    private static final int MAX_READ_SIZE = 2 * 1024 * 1024;
+
+    private final DirectSegmentAccess segment;
+    private final long startOffset;
+    private final int maxLength;
+    private final boolean shouldClear;
+    private final Duration fetchTimeout;
+    private final EntrySerializer entrySerializer;
+    private final ConvertResult<T> resultConverter;
+    private final Executor executor;
+    @GuardedBy("this")
+    private Iterator<Map.Entry<EntryIteratorState, TableEntry>> currentEntry = null;
+    @GuardedBy("this")
+    private long currentBatchOffset;
+    //endregion
+
+    private TableEntryIterator<T> setCurrentBatchOffset(long currentBatchOffset) {
+        this.currentBatchOffset = currentBatchOffset;
+        return this;
+    }
+
+    //region AsyncIterator Implementation
+
+    @Override
+    public CompletableFuture<T> getNext() {
+        // Verify no other call to getNext() is currently executing.
+        return getNextEntry()
+                .thenCompose(entry -> {
+                    if (entry == null) {
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        return this.resultConverter.apply(entry);
+                    }
+                });
+    }
+
+    public boolean endOfSegment() {
+        return this.currentBatchOffset >= (startOffset + maxLength);
+    }
+
+    private CompletableFuture<Map.Entry<EntryIteratorState, TableEntry>> getNextEntry() {
+        val entry = getNextEntryFromBatch();
+        if (entry != null) {
+            return CompletableFuture.completedFuture(entry);
+        }
+
+        return fetchNextTableEntriesBatch().thenApply(val -> getNextEntryFromBatch());
+    }
+
+    private synchronized Map.Entry<EntryIteratorState, TableEntry> getNextEntryFromBatch() {
+        if (this.currentEntry != null) {
+            Map.Entry<EntryIteratorState, TableEntry> next = this.currentEntry.next();
```

Comment:
Tip: use `val` to avoid having to write long type names (i.e `val next = this.currentEntry.next()`)